src/session.ts

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0
+"use strict";
+
+import { ethers, Signer } from "ethers";
+
+import { TxReceipt, BalanceProof } from "#erdstall/api/responses";
+import { Transfer, Mint, ExitRequest, TradeOffer, Trade } from "#erdstall/api/transactions";
+import { EnclaveWriter } from "#erdstall/enclave";
+import { Address, LedgerWriter } from "#erdstall/ledger";
+import { Assets } from "#erdstall/ledger/assets";
+import { Uint256 } from "#erdstall/api/util";
+import { Stages } from "#erdstall/utils";
+import { ErdstallSession } from "./erdstall";
+import { Client } from "./client";
+
+export const ErrUnitialisedClient = new Error("client unitialised");
+
+export class Session extends Client implements ErdstallSession {
+	readonly address: Address;
+	private nonce: bigint;
+	private readonly enclaveWriter: EnclaveWriter;
+	private readonly signer: Signer;
+
+
+	static dial(
+		address: Address,
+		signer: Signer,
+		operatorAddress: URL
+	): Session {
+		return new Session(address, new Client(signer, operatorAddress));
+	}
+
+	// Consumes `client`. `signer` is only needed if the client has no signer.
+	// `client` must have an EnclaveWriter.
+	constructor(address: Address, client: Client, signer?: Signer) {
+		super(client);
+		if((this.enclaveConn as EnclaveWriter).isEnclaveWriter !== undefined)
+			this.enclaveWriter = this.enclaveConn as EnclaveWriter;
+		else
+			throw new Error("enclave connection cannot be promoted to EnclaveWriter");
+
+		this.address = address;
+		this.nonce = 1n;
+
+		if(signer)
+			this.signer = signer;
+		else if(this.provider instanceof Signer)
+			this.signer = this.provider as Signer;
+		else
+			throw new Error("client provider cannot be promoted to Signer, no Signer provided");
+		this.provider = this.signer;
+
+		this.enclaveWriter.once("error", () => { this.updateNonce(); });
+	}
+
+	private async nextNonce(): Promise<bigint> {
+		if (this.nonce === 1n) {
```

Comment:
why don't we start at 0?

Reply:
Great catch. I changed it to start as 0, and also fixed a potential race in the updateNonce function when sending multiple transactions concurrently.